File: Assets/Animation.lua
local ____lualib = require("lualib_bundle")
local __TS__Class = ____lualib.__TS__Class
local __TS__SetDescriptor = ____lualib.__TS__SetDescriptor
local ____exports = {}
____exports.default = __TS__Class()
local Animation = ____exports.default
Animation.name = "Animation"
function Animation.prototype.____constructor(self, offsetX, offsetY, asset)
    self.animDuration = 0.15
    self.animElapsed = 0
    self.animOffsetX = 0
    self.animOffsetY = 0
    self.animTargetOffsetX = 0
    self.animTargetOffsetY = 0
    self.isAnimating = false
    self.originalX = 0
    self.originalY = 0
    self.originalX = asset.x
    self.originalY = asset.y
    self.animTargetOffsetX = offsetX
    self.animTargetOffsetY = offsetY
    self.animElapsed = 0
    self.isAnimating = true
    self.asset = asset
end
function Animation.prototype.updateAnimation(self, deltaTime)
    if not self.isAnimating then
        return
    end
    self.animElapsed = self.animElapsed + deltaTime
    if self.animElapsed >= self.animDuration then
        self.animElapsed = self.animDuration
        self.isAnimating = false
    end
    local progress = self.animElapsed / self.animDuration
    self.animOffsetX = self.animTargetOffsetX * progress
    self.animOffsetY = self.animTargetOffsetY * progress
    self.asset.y = self.originalY + self.animOffsetY
    self.asset.x = self.originalX + self.animOffsetX
end
__TS__SetDescriptor(
    Animation.prototype,
    "isFinished",
    {get = function(self)
        return not self.isAnimating
    end},
    true
)
return ____exports

File: Assets/Animation.ts
import Asset from "./Asset"

export default class Animation {
    animDuration: number = 0.15  // seconds
    animElapsed: number = 0
    animOffsetX: number = 0  // Current animation offset
    animOffsetY: number = 0  // Current animation offset
    animTargetOffsetX: number = 0  // Target animation offset
    animTargetOffsetY: number = 0  // Target animation offset (e.g., -20 for up)
    isAnimating: boolean = false
    originalX: number = 0
    originalY: number = 0   
    asset: Asset 

    constructor(offsetX: number, offsetY: number, asset: Asset) {
        this.originalX = asset.x
        this.originalY = asset.y
        this.animTargetOffsetX = offsetX
        this.animTargetOffsetY = offsetY
        this.animElapsed = 0
        this.isAnimating = true
        this.asset = asset
    }

    updateAnimation(deltaTime: number): void {
        if (!this.isAnimating) {
            return
        }

        this.animElapsed += deltaTime
        
        if (this.animElapsed >= this.animDuration) {
            // Animation complete
            this.animElapsed = this.animDuration
            this.isAnimating = false
        }

        // Interpolate animation offset
        const progress = this.animElapsed / this.animDuration
        this.animOffsetX = this.animTargetOffsetX * progress
        this.animOffsetY = this.animTargetOffsetY * progress
        
        this.asset.y = this.originalY + this.animOffsetY
        this.asset.x = this.originalX + this.animOffsetX
    }

    get isFinished(): boolean {
        return !this.isAnimating
    }
}